### DISCUSSION TITLE
Are there plans to get this merged into Home Assistant

### DISCUSSION THREAD

**Discussion by lanrat** [custom-components/ble_monitor#224] [General]
Are there plans to eventually get this integration merged into Home Assistant and possibly replace the `mitemp_bt` integration?

And if so, what would it take to do that?

**Comment by Ernst79** [comment]
Lets start what is needed to get integrated in Home Assistant. 

The most important thing that need to be done are
- To remove YAML configuration, as this is [not allowed](https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md#decision) for new integrations anymore
- To [add tests](https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations)
- All [communication to external devices](https://developers.home-assistant.io/docs/development_checklist) or services must be wrapped in an external Python library hosted on pypi. This means that our modified aioblescan.py has to move to a pypi package
- Fulfill all other [requirements](https://developers.home-assistant.io/docs/creating_component_code_review) for new platforms

The pro's of creating an official integration
- Easier to find/install
- More users
- More developers that will contribute?
- Better code quality (after review of the HA devs)
- Automatic testing of the code

The cons of creating an official integration
- Losing YAML configuration
- Beta testing will be a bit more difficult, as we cannot easily make beta versions. 
- Changes will be more time consuming, due to the process of reviews, etc. 

So, Yes, I agree it would be good to eventually get the integration in Home Assistant. However, I think time (and for me personally proper python experience) is holding us back from doing it. Developing tests is not the most grateful work to do, removing YAML also won't be appreciated by many (although I'm not against personally).

**Reply by spbrogan** [reply]
I am also curious about this and/or interested in trying to merge the numerous similar HACS projects.  I have created my own bluetooth sensor (https://github.com/spbrogan/sensor.mopeka_pro_check) and my own bluetooth pypi sensor lib (https://github.com/spbrogan/mopeka_pro_check).  These were created to support a BLE LP tank sensor.  According to the project I derived from (https://github.com/Home-Is-Where-You-Hang-Your-Hack/sensor.goveetemp_bt_hci) it is based on an older version of this project.  

My version is very specific to the tank sensor but I did review the HA requirements before starting.  I split up the external communication into its own library available thru pypi.  I created a sensor object in the library and then in the HA code I just wrapped that into a HA specific sensor.  I also have a few unit tests.  

Anyway I am curious if there is any interest in joining these three and maybe creating a solution that is architected to cleanly be extended to new sensors as well as separate the logic between HA and BLE.

**Reply by Ernst79** [reply]
yes, I'm certainly interested in merging several components. Main problem is that I'm currently all doing it by my own, so help is really needed if you want to expand BLE monitor with more sensors. I'm just doing this as a hobby, I'm not an programmer, just trying stuff by trial and error. :-)

Govee is indeed based on an old version of BLE monitor, at the time it was still called mitemp_bt. But it should be possible to merge this, if there is someone who wants to do some testing and generates data (e.g. a HCI dump). The same applies for your sensors.

So, yes, if you want to help, I'm certainly willing to merge components. 

The parsing of the current sensors is at the end of the `__init__.py` file (currently ATC, qingping and Xiaomi MiBeacon sensor). This should move to an external pypi package preferably, in combination with aioblescan_ext.py.

So, if you decide to join forces, please open a new issue where we discuss how to go forward and what we want to achieve. We can than work on that together.
